From 3e41c8b8cfd2c21ad9238569f7872970115a6a8b Mon Sep 17 00:00:00 2001
From: Christian Walther <walther@indel.ch>
Date: Thu, 27 Nov 2014 12:00:49 +0100
Subject: [PATCH] Add wxEXPAND_ROWS flag to wxWrapSizer.

It causes the sizer to expand its rows to take up all available space in the minor direction rather than keeping them at their minimum size.
---
 include/wx/wrapsizer.h   |    2 ++
 interface/wx/wrapsizer.h |   18 +++++++++++++-----
 src/common/wrapsizer.cpp |    2 +-
 src/xrc/xh_sizer.cpp     |    1 +
 4 files changed, 17 insertions(+), 6 deletions(-)

diff --git a/include/wx/wrapsizer.h b/include/wx/wrapsizer.h
index 2154de3..6b7562c 100644
--- a/include/wx/wrapsizer.h
+++ b/include/wx/wrapsizer.h
@@ -18,6 +18,8 @@ enum
     wxEXTEND_LAST_ON_EACH_LINE = 1,
     // don't leave spacers in the beginning of a new row
     wxREMOVE_LEADING_SPACES = 2,
+    // expand rows to take up all available space in the minor direction
+    wxEXPAND_ROWS = 4,
 
     wxWRAPSIZER_DEFAULT_FLAGS = wxEXTEND_LAST_ON_EACH_LINE |
                                 wxREMOVE_LEADING_SPACES
diff --git a/interface/wx/wrapsizer.h b/interface/wx/wrapsizer.h
index 61bc160..9a0f341 100644
--- a/interface/wx/wrapsizer.h
+++ b/interface/wx/wrapsizer.h
@@ -10,6 +10,7 @@ enum
 {
     wxEXTEND_LAST_ON_EACH_LINE,
     wxREMOVE_LEADING_SPACES,
+    wxEXPAND_ROWS,
     wxWRAPSIZER_DEFAULT_FLAGS
 };
 
@@ -37,11 +38,18 @@ public:
 
         @a orient determines the primary direction of the sizer (the most common
         case being @c wxHORIZONTAL). The flags parameter can be a combination of
-        the values @c wxEXTEND_LAST_ON_EACH_LINE which will cause the last item
-        on each line to use any remaining space on that line and @c wxREMOVE_LEADING_SPACES
-        which removes any spacer elements from the beginning of a row.
+        the following values:
+        <dl>
+        <dt>@c wxEXTEND_LAST_ON_EACH_LINE</dt><dd>cause the last item on each
+        line to use any remaining space on that line</dd>
+        <dt>@c wxREMOVE_LEADING_SPACES</dt><dd>remove any spacer elements from
+        the beginning of a row</dd>
+        <dt>@c wxEXPAND_ROWS</dt><dd>expand rows to take up all available space
+        in the minor direction (default is to keep them at minimum size)</dd>
+        </dl>
 
-        Both of these flags are on by default.
+        @c wxEXTEND_LAST_ON_EACH_LINE and @c wxREMOVE_LEADING_SPACES are on by
+        default.
     */
     wxWrapSizer(int orient = wxHORIZONTAL,
                 int flags = wxWRAPSIZER_DEFAULT_FLAGS);
diff --git a/src/common/wrapsizer.cpp b/src/common/wrapsizer.cpp
index e1b112b..728f29b 100644
--- a/src/common/wrapsizer.cpp
+++ b/src/common/wrapsizer.cpp
@@ -127,7 +127,7 @@ wxSizer *wxWrapSizer::GetRowSizer(size_t n)
         return rows[n]->GetSizer();
 
     wxSizer * const sizer = new wxBoxSizer(GetOrientation());
-    m_rows.Add(sizer, wxSizerFlags().Expand());
+    m_rows.Add(sizer, wxSizerFlags((m_flags & wxEXPAND_ROWS) ? 1 : 0).Expand());
     return sizer;
 }
 
diff --git a/src/xrc/xh_sizer.cpp b/src/xrc/xh_sizer.cpp
index ab08f86..bbc3069 100644
--- a/src/xrc/xh_sizer.cpp
+++ b/src/xrc/xh_sizer.cpp
@@ -89,6 +89,7 @@ wxSizerXmlHandler::wxSizerXmlHandler()
     // wxWrapSizer-specific flags
     XRC_ADD_STYLE(wxEXTEND_LAST_ON_EACH_LINE);
     XRC_ADD_STYLE(wxREMOVE_LEADING_SPACES);
+    XRC_ADD_STYLE(wxEXPAND_ROWS);
 }
 
 
-- 
1.7.5.4

